### PR TITLE
Fix linking of shared/static libs with static libs

### PR DESCRIPTION
**Pull request by nirbheek**
This commit contains the following fixes:

1. When a shared library A does `link_with:` to static library B, the
   parts of B used by A will be added to A, and so we don't need to
   return B in A.get_dependencies() for targets that link to A. This
   already is the behaviour when a shared library A does `link_whole:`
   on B.

2. In situation (1), when generating a pkg-config file for A, we must
   also not add B to Libs.private for A. This already is the behaviour
   when a shared library A does `link_whole:` on B.

3. When a static library A does `link_whole:` to static library B, we
   must add the objects in B to A.

4. When a static library A does `link_with:` to static library B, and
   B is not installed (which makes it an internal static library), we
   must add the objects in B to A, otherwise nothing can use A.

5. In situation (4), when generating a pkg-config file for A, we must
   also not add B to Libs.private for A.

All these situations are tested by the unit test added in this commit.

Closes https://github.com/mesonbuild/meson/issues/3934
Closes https://github.com/mesonbuild/meson/issues/3937

### CODE REVIEW

**Comment by nirbheek** [comment]
Ready for merging after review. This is a pretty important issue, without which static libraries in glib and gstreamer are completely unusable.

**Comment by nirbheek** [comment]
Tests fixed.

**Comment by jpakkane** [comment]
Has this been tested with gstreamer et al?

**Comment by nirbheek** [comment]
I've tested it with gstreamer and glib. I want to test more, but was focussing on the stable release.

**Comment by jpakkane** [comment]
Is this ready to merge now?

**Comment by nirbheek** [comment]
I'd like https://github.com/mesonbuild/meson/pull/4061 and https://github.com/mesonbuild/meson/pull/3894 merged first, to avoid conflicts and to make it easier for me to backport those to 0.47.

**Comment by textshell** [comment]
Sorry for just now noticing this.

Do we really want to change the output of the build depending on how a library is installed? Maybe a library is installed using a install script? Maybe a project doesn't even use meson installation support?

Point 4 and 5 seem like very implicit magic to me. At the very most this needs clear documentation for our users.

**Comment by jpakkane** [comment]
It seems more straightforward that if you "link whole" a static library to another, the end result should be the same as when plain linking a static library to another. That is: it should be "link whole"d into the final target that uses it.

**Comment by nirbheek** [comment]
> Do we really want to change the output of the build depending on how a library is installed? Maybe a library is installed using a install script? Maybe a project doesn't even use meson installation support?

... this is how Autotools already works. See: https://github.com/mesonbuild/meson/issues/3937

> Point 4 and 5 seem like very implicit magic to me. At the very least this needs clear documentation for our users.

Point 4 is also how Autotools already works, and point 5 is how people expect pkg-config files to work.

**Comment by jpakkane** [comment]
The fact that autotools does X is a very strong recommendation to do the exact opposite, regardless of the value of X.

**Comment by nirbheek** [comment]
> The fact that autotools does X is a very strong recommendation to do the exact opposite, regardless of the value of X.

Is that why we copy Autotools behaviour in so many ways ranging from library names to option names? 😄 

In any case, my point was that this is not unexpected behaviour, it's what users already expect. The reason why we haven't seen this yet is because glib and gstreamer are still shipping with Autotools. Without these patches the static libraries in those projects are missing a lot of objects.

**Comment by textshell** [comment]
> It seems more straightforward that if you "link whole" a static library to another, the end result should be the same as when plain linking a static library to another. That is: it should be "link whole"d into the final target that uses it.

I'm not questioning the how "link whole" should work. I'm questioning that a normal link starts behaving more like a "link whole" depending on if the linked library is set to be installed. 

I have to admit, that i just now notices this PR so i couldn't yet actually grok the code changes.

But this really seems like too implicit having it tied to the install argument. Users will not expect that to change how the actual libraries are build. They might even just add it when everything works. Which is does not, because they have not added it yet. 

> Point 4 is also how Autotools already works

This i certainly an non argument in meson land. Or even an indication to think really hard if this is something that really makes sense or just yet another quirk.

**Comment by textshell** [comment]
> In any case, my point was that this is not unexpected behaviour, it's what users already expect. The reason why we haven't seen this yet is because glib and gstreamer are still shipping with Autotools. Without these patches the static libraries in those projects are missing a lot of objects.

I would certainly not expect this at all. I'd rather we force users to explicitly label internal only libraries than to tied it to something that surely is related but not really that closely. In my book "explicit is better than implicit" as long as it's not so much of a burden to make it impractical.

We do have choices here to make the meson builds for for the affected libraries.

**Comment by jpakkane** [comment]
So something like having a kwarg like `link_with_internal`? Would that work or would that get too confusing?

**Comment by jpakkane** [comment]
Then we could even verify and mandate that all those kinds of libraries must be static.

**Comment by textshell** [comment]
> Then we could even verify and mandate that all those kinds of libraries must be static.

Yes, i was just doing to ask what happens when that library is shared. I think erroring out might be a good start actually.

Now remind me, what is the difference between link_whole and link_with_internal?

For shared libraries? Seems like this only links in symbols (with function gc) / objects (without function gc) from the static lib that are needed?

For static libraries? Nothing? Could we later improve that to more closely match the shared case?

**Comment by jpakkane** [comment]
> For static libraries? Nothing? 

The only difference I can see is that it produces smaller binaries because the linker can strip out unneeded code.

**Comment by textshell** [comment]
> > For static libraries? Nothing?

> The only difference I can see is that it produces smaller binaries because the linker can strip out unneeded code.

I think the current code does not do that, but i think we could define meson's semantics in a way that this could be added later.

On to the bike-shedding. Is `link_with_internal` the right name for the kwarg?
* internal seems to imply something is hidden from the outside. In this case it would be the dependency but not the actual content. (modulo -fdefault-visibility as usual for shared libraries)
* On the other hand this type of linking is some kind of inclusion of the (needed) objects, so that might be an option too.
* Something else?

**Comment by textshell** [comment]
Do we need something like link_with_internal for use with declare_dependency too?

I think when used with a both_libraries return object, this needs to know to take the static library (not the default shared one)

**Comment by nirbheek** [comment]
I think the behaviour implemented by this PR is absolutely the expected behaviour for people who have a setup where they `link_with:` a helper static library into their shared library.  As shown by the build files written in glib and gstreamer, this is how people expected meson to work. We at gstreamer noticed it first because we started deploying meson for static-only builds intended for iOS and Android.

*Of course* you want the objects from the helper static library included into the shared library. Nothing else makes sense, unless of course you're installing the static library and then you want the `-lfoo` arg in the pkg-config file so users know that they should link to the static library.

This is different from linking to a shared library because that adds an entry to `DT_NEEDED`, so the linker will search for that shared library automatically. That doesn't work for static libraries.

One theoretical problem here is if the user is installing the static library through some other means, such as an install script. For that we could add a kwarg to the library to specify that it's not installed (and hence internal).

I don't think it makes sense to define the relationship through a new `link_*:` kwarg. That will only be confusing for people who expect "link to static library" to work as-is. I also think the property resides on the static library, not in the relationship between the static library and its consumer. I doubt there's a case where you want to treat a static library as internal in one link and external in another.

**Comment by jpakkane** [comment]
Just as a clarification: this functionality (or something like it) is needed and we should provide for it. Maybe we will do what Autotools does, but we should not be beholden to it. If we can come up with an API that is better (possibly more explicit, easier for newcomers, what have you) then we should use that instead. This is a major change in behaviour so it is worth thinking about before committing (but not endlessly bikeshedding, obviously).

**Comment by textshell** [comment]
Having the information explicit on the internal library by a kwarg like "internal" or something else was my first impulse. So i think that is an alternative option to 'link_with_internal'.

Just thinking out loud, these internal libraries are not really much of libraries, but more collections of objects for later reuse that have independent compiler flags (which allows them to be reused). We also have constant requests for having independent compiler flags every once in a while and users are put of by telling them to use a library because of some connotations of that word. Maybe there is something common struggling to be discovered. Something like a "part". Or maybe not^^

**Comment by nirbheek** [comment]
I think it's a good idea to allow people to override the default of interpreting the `install:` kwarg to determine whether the library is internal or not. 

One important deadline we have for this is that glib is planning on dropping Autotools entirely after their upcoming 2.58.0 release, and it would be good to have 0.48 out with this PR before that.

I also have the [macOS `compatibility_version` issue](https://github.com/mesonbuild/meson/issues/3555) on my list for targeting that.

**Comment by xclaesse** [comment]
I was wondering, with this PR, is there any difference between `link_with:` and `link_whole` with an uninstalled static library? And if there is no difference (?) why can't GLib/GStreamer just use `link_whole` already? Sorry if that has been answered already above, I'm jumping into the discussion here.

**Comment by xclaesse** [comment]
Now that I think about it, I already faced this problem in glib-networking, my workaround was to install the static library, but I guess that will have to be revisited once we have a fix in meson. See libgiopkcs11 in https://gitlab.gnome.org/GNOME/glib-networking/blob/master/tls/pkcs11/meson.build.

**Comment by nirbheek** [comment]
> I was wondering, with this PR, is there any difference between link_with: and link_whole with an uninstalled static library?

The difference is that `link_whole:` will add all the objects of that static library to the shared library (wholesale), while `link_with:` will only add those objects that are needed. If you use `--gc-sections`, the unused sections will also be removed from those objects.

> Now that I think about it, I already faced this problem in glib-networking, my workaround was to install the static library

Yes, this should no longer be needed once this PR lands.

**Comment by nirbheek** [comment]
> Now that I think about it, I already faced this problem in glib-networking

FWIW, glib-networking should use `link_whole:` already on the `giopkcs11` static library since the use-case is "static library that is just a holder for objects that will go into the main library", and you *don't* want to drop unused sections. 

It's different for glib because the use-case there is "these symbols/libraries were not found on the system, so I have an internal copy for that case", and you *do* want to drop unused sections.

**Comment by nirbheek** [comment]
> FWIW, glib-networking should use link_whole: already on the giopkcs11 static library

... of course this won't work in the static case because we were doing nothing when a static library did a `link_whole:` on a static library, but it will be fixed with this PR.

**Comment by nirbheek** [comment]
So, what do we do here? Add a kwarg `internal_helper:` or similar to `static_library()` to disable the automatic-detection based on the `install:` kwarg? I would consider this PR a blocker for 0.48 because without this glib can't really move from Autotools to Meson.

**Comment by jpakkane** [comment]
> When a shared library A does link_with: to static library B, the parts of B used by A will be added to A, and so we don't need to return B in A.get_dependencies() for targets that link to A.

But what if B was linked against shared library C? We should return that for `get_dependencies` of A, right? Does the current code do that?

I have been thinking about this behaviour for a while now. Here are some questions that came up:

- can we do this now and add the `link_with_internal_something` later if an actual need for it arises?
- if we go with the current behaviour it needs to be thoroughly documented, because this behaviour is very unexpected to people who have not encountered it before (which, I'm guessing, is everyone not proficient in Autotools)

**Comment by nirbheek** [comment]
> But what if B was linked against shared library C? We should return that for `get_dependencies` of A, right? Does the current code do that?

Yes, it does, and we have tests for it too; see `common/150 recursive linking`:

`150 recursive linking/3rdorderdeps/meson.build`
```
# Permutate all combinations of shared and static libraries up to three levels
# executable -> shared -> static -> shared (etc)
foreach dep2 : ['sh', 'st']
  foreach dep1 : ['sh', 'st']
    foreach libtype : ['sh', 'st']
...
```

So it tests all possible combinations, and everything beyond a third-order dependency is composed from 1st, 2nd, and 3rd order deps.

> * can we do this now and add the `link_with_internal_something` later if an actual need for it arises?

This would be my preferred solution. I also don't mind adding a property on `static_library()` to make the role of that library explicit (in case the user does not want us to infer it from the `install:` kwarg) since the cost of doing that is minimal.

However, I think the API cost of adding a new `link_with_foo` type is much larger than just a property on `static_library()` though, so that should only be added after someone actually has a good use-case for it.

> * if we go with the current behaviour it needs to be thoroughly documented, because this behaviour is very unexpected to people who have not encountered it before

Sure, more documentation is always good. I can add a release notes entry and document it under both `link_with:` and under `install:` for `static_library()`.

**Comment by textshell** [comment]
I'm still strongly in favor of something explicit. I don't have a strong opinion about if that is in the library creation or in the linkage kwargs.

Placing it with the helper library likely creates less api surface (linkage kwargs might need to also be added to internal dependency declarations etc in the long run) but it feels a bit like it deserves an separate name apart from static_library. Which of course would again be more api surface in a way.

File: run_unittests.py [hunk 1]
```diff
@@ -2744,6 +2752,81 @@ def test_buildtype_setting(self):
...
         self.assertEqual(opts['optimization'], '0')
 
+    def test_static_and_shared_library_usability(self):
+        '''
+        Test that static and shared libraries with various kinds of static
+        library internal dependencies are usable after installation, and that
+        the pkg-config files generated for such libraries have the correct
+        Libs: and Libs.private: lines.
+        '''
+        env = Environment('', '', get_fake_options(''))
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

[F821] undefined name 'Environment'

**Comment by nirbheek** [comment]
Rebased on top of master. We need to decide soon whether we want this in with a kwarg on `static_library()` or not, because otherwise we won't have enough time before the release for people to test it.

Note that this does not change uninstalled usage of projects, so the only use-case that we have seen for a kwarg on `static_library()` is if people are using a post-install script for installing a static library that is also linked into another static or shared library.

**Comment by Cogitri** [comment]
It seems like this patch breaks mesa's build (tested with 18.2.0&scm):
https://paste.pound-python.org/show/ta2EK7BmhKx8F734LeFJ/

mesa links the shared libGL against their static libglx, see: https://cgit.freedesktop.org/mesa/mesa/tree/src/glx/meson.build

**Comment by nirbheek** [comment]
Hum, that's a bug in this PR. Somehow we're adding the objects from `libmesa_util` twice. Fixing.

**Comment by jpakkane** [comment]
Thinking about this further, it is actually possible to do this without any change in Meson with something like this:

```meson
if building_static
  dep_objs = internal_library.extract_all_objects()
  internal_dep = [] # or with compilation options
else
  dep_objs = []
  internal_dep = internal_dep # Yes, this is a no-op
endif

library('glib', ...., dep_objs, dependencies : [internal_dep, ...])
```

So maybe we don't need this at all? Or maybe we could have a top level function to convert static library deps to "internal object" deps? Or maybe that could be a method on the dependency object like `partial_dependency`?

This has the added nicety that the thing you are trying to achieve is visible in the build definitions so even people who don't know about this feature can fairly easily tell what is happening. What do you think?

**Comment by textshell** [comment]
Is that if with building_static really a real use case? These parts are not really configurable they are always parts that are static and only to be used to avoid recompiling sources for multiple target and/or allowing different compiler flags.

If we want something more semantic i still thing what i said her https://github.com/mesonbuild/meson/pull/3939#issuecomment-414740967 is a good idea to think about. Have a new target type that is "just a collection of compiled source with independent configuration" that defines a clear concept of a group of sources with independent settings. We from time to time get request for various fiddly things that could all find a home there.

**Comment by jpakkane** [comment]
Can we release 0.48.0 without this or is this important enough to delay the release?

**Comment by nirbheek** [comment]
If we're ok with possibly releasing 0.49 early with this patch for glib, I think it's better to punt this to the next release at this point. We won't get enough testing for it anyway, and I'll have less time this month and the next to work on the stable release.

Current master has a lot of features that are already important for glib and it'll be good to get them out.

**Comment by smcv** [comment]
> these internal libraries are not really much of libraries, but more collections of objects for later reuse that have independent compiler flags (which allows them to be reused)

For what it's worth, the Autotools jargon for this is a "convenience library". That isn't a very clear or obvious term, so it would be fine to choose a better name for the equivalent Meson feature, but it might be useful to mention "this is the same thing as an Autotools convenience library" in documentation.

**Comment by xclaesse** [comment]
@ebassi fwiw, I'm 100% for implicitly consider not-installed static_library() as internal. Was saying that just in case people insist in making it somehow explicit.

**Comment by Cogitri** [comment]
Although I'm not a meson dev I do use meson a lot and personally I'd +1 on making it implicit. One of meson's strong points is how easy it is to use.

**Comment by ePirat** [comment]
Recently [an issue](https://code.videolan.org/videolan/dav1d/issues/44) was raised for dav1d, a new project with a somewhat complex structure of helper libraries.

While this issue fixes the mentioned issue partially, it seems that it does not work properly for generated objects which have to be added as sources to the library. These objects are just missing in the final library, which breaks a lot of helper-library use-cases in subtle ways that are very hard to figure out. It seems `extract_all_objects()` is affected by the same problem, see #2479.

File: mesonbuild/build.py [hunk 2]
```diff
@@ -856,22 +856,43 @@ def get_outputs(self):
...
+        for t in self.link_whole_targets:
+            if t in transitive_deps or t in exclude:
+                continue
+            if not internal and t.is_internal():
+                continue
+            # self.link_whole_targets are not included by default here because
+            # the objects from those will already be in the library. They are
+            # only needed while generating backend (ninja) target dependencies.
+            if link_whole:
+                transitive_deps.append(t)
```
Comment (xclaesse) [inline mesonbuild/build.py]: Here we have 2 loops where the only difference is those 2 lines. Would be nice to find a way to merge them. The condition could be for example `if link_whole or t not in self.link_whole_targets` ?

File: mesonbuild/build.py [hunk 3]
```diff
@@ -958,7 +979,17 @@ def link(self, target):
                 raise InvalidArguments(msg)
             if self.is_cross != t.is_cross:
                 raise InvalidArguments('Tried to mix cross built and native libraries in target {!r}'.format(self.name))
-            self.link_targets.append(t)
+            # When linking to a static library that's not installed, we
+            # transparently add that target's objects to ourselves.
+            # Static libraries that are installed will either be linked through
+            # self.link_targets or using the pkg-config file.
+            if isinstance(self, StaticLibrary) and isinstance(t, StaticLibrary) and not t.need_install:
```
Comment (xclaesse) [inline mesonbuild/build.py]: `if isinstance(self, StaticLibrary) and t.is_internal()` since we have that nice small helper.

File: mesonbuild/build.py [hunk 4]
```diff
@@ -958,7 +979,17 @@ def link(self, target):
...
+                self.objects.append(t.extract_all_objects())
+                # Add internal and external deps
+                self.external_deps += t.external_deps
+                self.link_targets += t.link_targets
```
Comment (xclaesse) [inline mesonbuild/build.py]: Don't we need to recurse here? In the case an internal library link to another internal library.

File: mesonbuild/build.py [hunk 5]
```diff
@@ -970,7 +1001,15 @@ def link_whole(self, target):
...
             if self.is_cross != t.is_cross:
                 raise InvalidArguments('Tried to mix cross built and native libraries in target {!r}'.format(self.name))
-            self.link_whole_targets.append(t)
+            # When we're a static library and we link_whole: to another static
+            # library, we need to add that target's objects to ourselves.
+            if isinstance(self, StaticLibrary):
+                self.objects.append(t.extract_all_objects())
+                # Add internal and external deps
+                self.external_deps += t.external_deps
+                self.link_targets += t.link_targets
```
Comment (xclaesse) [inline mesonbuild/build.py]: Same here, we could need to recurse I think.

File: test cases/common/143 C and CPP link/meson.build [hunk 6]
```diff
@@ -15,8 +15,8 @@
 project('C and C++ static link test', ['c', 'cpp'])
 
 # Verify that adding link arguments works.
-add_global_link_arguments('', language : 'c')
-add_project_link_arguments('', language : 'c')
+add_global_link_arguments('-DMESON_UNUSED', language : 'c')
+add_project_link_arguments('-DMESON_UNUSED', language : 'c')
```
Comment (xclaesse) [inline test cases/common/143 C and CPP link/meson.build]: This seems unrelated to this PR.

File: run_unittests.py [hunk 7]
```diff
@@ -2744,6 +2752,81 @@ def test_buildtype_setting(self):
...
+        else:
+            static_args = ''
+            libtypes = ('static', 'shared', 'both')
+        # Test
+        for libtype in libtypes:
+            oldprefix = self.prefix
+            # Install external library so we can find it
+            testdir = os.path.join(self.unit_test_dir, '35 both library usability', 'provider')
+            # install into installdir without using DESTDIR
+            installdir = self.installdir
```
Comment (xclaesse) [inline run_unittests.py]: That's a bit hackish, you could test -uninstalled.pc files here, when that PR gets in. But I guess that's fine for a unit test.

**Comment by ocrete** [comment]
This fixed libnice's build, and is a blocker to move libnice to meson.

**Comment by xclaesse** [comment]
@nirbheek ping?

**Comment by nirbheek** [comment]
I'll revise this PR next week.

**Comment by xclaesse** [comment]
> I'll revise this PR next week.

ping?

**Comment by nirbheek** [comment]
I don't think it's a good idea to push this in last-minute, but yes I will find some time for this in the next couple of days so we can get it in after the release.

**Comment by Cogitri** [comment]
So...very sorry to ask for an ETA here, but I suppose this will be fixed soon-ish (not quite sure when 0.50.0 is scheduled to release). Glib 2.60.0 has been released and can't be used on any distro which needs its static libs due to them dropping autotools and this not being fixed yet which sort of blocks GNOME 3.32 at least on Void Linux :/

**Comment by ocrete** [comment]
Just one comment here, I don't think this should be specific to internal "convinience" libraries. I'd like to be able to have the same behaviour when linking in static libraries built with some other tool into my meson-built library.

**Comment by aduskett** [comment]
Where are we on this? I have just tested this patch series applied to meson .050.1 and GStreamer 1.16.0 with a static build using buildroot and it fixes all the linking issues. It would be nice to not have to include a 406 line patch in BuildRoot

**Comment by jpakkane** [comment]
Well this one :crossed_swords: conflicts currently so it needs a rebase at least. @nirbheek what is your opinion on this now? Is it ready for merging or does it still need work?

**Comment by aduskett** [comment]
For what it's worth, I did apply this merge request to the latest release and it does fix static linking with gstreamer 1.16.0 and glib 2.60.2

**Comment by nirbheek** [comment]
> @nirbheek what is your opinion on this now? Is it ready for merging or does it still need work?

I don't think this PR can go in as-is anymore because projects have added workarounds that will cause errors now. I think the last time I checked, Mesa was one such project.

One part of this can go in, and that's the `link_whole:` behaviour with static libraries, point (3) in the PR description and (5) if it is still needed. The others are harder to get right. We'll have to break compatibility to add them, so we'll need a solid plan there.

**Comment by eli-schwartz** [comment]
Similar issue: #4984

Maybe A `link_with/link_whole: B` does not need to return B as a dependency, but sometimes it does need to return B's dependencies.

**Comment by xclaesse** [comment]
> Glib 2.60.0 has been released and can't be used on any distro which needs its static libs

I don't think that's true, Glib merged a workaround for this and I even wrote a CI test that ensure that installed glib can be used to static link an app. But if you have a specific issue, please report.

**Comment by Cogitri** [comment]
> I don't think that's true, Glib merged a workaround for this and I even wrote a CI test that ensure that installed glib can be used to static link an app. But if you have a specific issue, please report.

That may very well be the case now, it definitively was broken for Void Linux with GLib 2.60.0, but there have been some GLib release since that.

File: mesonbuild/build.py (hunk 5 above)
Comment (xavierclaessens) [reply mesonbuild/build.py]: Revisiting a bit this PR, and I have a question here: Why do you need to `extract_all_objects()` in the link_whole case? Isn't that exactly what the linker is going to do when we'll do `--whole-archive` ? Also since you removed `self.link_whole_targets.append(t)`, as far as I understand `self.link_whole_targets` list will now always be empty.

Probably I'm missing something here, but it sounds like we're never going to use --whole-archive any more, and always just give all objects files instead. Is that on purpose?

Comment (xclaesse) [reply mesonbuild/build.py]: Oh, sorry, didn't see the 'else' case, so `link_whole_targets` is still used for shared libraries.

**Comment by xclaesse** [comment]
I made a new PR based on this one, based on the parts that I actually understand. Can everyone here give it a try in their project (especially Meson) and report any regression, please?

**Comment by xclaesse** [comment]
`3.`, `4.` and `5.` should be fixed by https://github.com/mesonbuild/meson/pull/5936.

>     1. When a shared library A does `link_with:` to static library B, the
>        parts of B used by A will be added to A, and so we don't need to
>        return B in A.get_dependencies() for targets that link to A. This
>        already is the behaviour when a shared library A does `link_whole:`
>        on B.

I don't think this is what we want here. If we omit B from get_dependencies(), won't it break ninja dep graph too? Also if we have a library C that link_with A, since A is shared library C.get_dependencies() never included B. We never linked C to B. So I think it was already working as expected.

>     2. In situation (1), when generating a pkg-config file for A, we must
>        also not add B to Libs.private for A. This already is the behaviour
>        when a shared library A does `link_whole:` on B.

pkgconfig generator is a bit more complicated. If for A you use shared_library() then it will never include `Libs.private` anyway. If it's both_libraries() then it does need to add B into `Libs.private` otherwise the static A won't work. So I think this has always been working as expected.

**Comment by xclaesse** [comment]
https://github.com/mesonbuild/meson/pull/5936 is now merged, I think we can close this PR since I think all valid cases have been covered already. Any remaining issue can be tackled in a new PR. Thanks everyone who participated.
